From 8b7dfb59ebee9d8c8041811c09f4ec1c37c07ee1 Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 11 Aug 2023 13:15:21 +0800
Subject: [PATCH] Add new rpc caller for Protected Storage in se_proxy SP

The new rpc caller is based on the rss comms acted as
conversion layer which is referenced from the implementation
in trusted-firmware-m [0] and MHU V3 driver acted as
transport layer. The conversion layer is used to
serialize and deserialize the messages between Trusted
Service and RSS for kronos platform.

0:
platform/ext/target/arm/rss/common/rss_comms/rss_comms_protocol.c
platform/ext/target/arm/rss/common/rss_comms/rss_comms_protocol_pointer_access.c
platform/ext/target/arm/rss/common/rss_comms/rss_comms_protocol_embed.c

https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../messaging/rss_comms/sp/component.cmake    |  24 +++
 .../rss_comms/sp/rss_comms_messenger.c        |  42 ++++
 .../rss_comms/sp/rss_comms_messenger.h        |  22 +++
 .../rss_comms/sp/rss_comms_messenger_api.h    |  23 +++
 .../messaging/rss_comms/sp/rss_comms_mhu.c    | 121 ++++++++++++
 .../messaging/rss_comms/sp/rss_comms_mhu.h    |  28 +++
 .../messaging/rss_comms/sp/rss_comms_virtio.c | 114 +++++++++++
 .../messaging/rss_comms/sp/rss_comms_virtio.h |  36 ++++
 .../rss_comms/caller/sp/rss_comms_caller.c    |  98 ++++++++++
 .../rss_comms/caller/sp/rss_comms_caller.h    |  22 +++
 .../rss_comms/caller/sp/rss_comms_protocol.c  |  74 +++++++
 .../rss_comms/caller/sp/rss_comms_protocol.h  |  69 +++++++
 .../caller/sp/rss_comms_protocol_embed.c      |  95 +++++++++
 .../caller/sp/rss_comms_protocol_embed.h      |  47 +++++
 .../sp/rss_comms_protocol_pointer_access.c    |  73 +++++++
 .../sp/rss_comms_protocol_pointer_access.h    |  44 +++++
 components/rpc/rss_comms/component.cmake      |  27 +++
 components/rpc/rss_comms/service_rss_comms.c  | 182 ++++++++++++++++++
 .../secure_storage_ipc/secure_storage_ipc.c   |   7 +-
 .../kronos-opteesp/default_se-proxy.dts.in    |  18 ++
 deployments/se-proxy/infra/kronos/infra.cmake |   9 +
 .../infra/kronos/service_proxy_factory.c      |  47 ++++-
 platform/drivers/arm/mhu_driver/mhu.h         |   1 +
 platform/drivers/arm/mhu_driver/mhu_v3_x.c    |  10 -
 platform/drivers/arm/mhu_driver/mhu_v3_x.h    |   2 -
 .../drivers/arm/mhu_driver/mhu_wrapper_v3_x.c |  23 ++-
 26 files changed, 1233 insertions(+), 25 deletions(-)
 create mode 100644 components/messaging/rss_comms/sp/component.cmake
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_messenger.c
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_messenger.h
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_messenger_api.h
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_mhu.c
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_mhu.h
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_virtio.c
 create mode 100644 components/messaging/rss_comms/sp/rss_comms_virtio.h
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_caller.c
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_caller.h
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol.c
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol.h
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.c
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.h
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.c
 create mode 100644 components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.h
 create mode 100644 components/rpc/rss_comms/component.cmake
 create mode 100644 components/rpc/rss_comms/service_rss_comms.c

diff --git a/components/messaging/rss_comms/sp/component.cmake b/components/messaging/rss_comms/sp/component.cmake
new file mode 100644
index 0000000..f797713
--- /dev/null
+++ b/components/messaging/rss_comms/sp/component.cmake
@@ -0,0 +1,24 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021-2023, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+if (NOT DEFINED TGT)
+	message(FATAL_ERROR "mandatory parameter TGT is not defined.")
+endif()
+
+set_property(TARGET ${TGT} APPEND PROPERTY PUBLIC_HEADER
+	"${CMAKE_CURRENT_LIST_DIR}/rss_comms_messenger_api.h"
+	)
+
+target_sources(${TGT} PRIVATE
+	"${CMAKE_CURRENT_LIST_DIR}/rss_comms_messenger.c"
+	"${CMAKE_CURRENT_LIST_DIR}/rss_comms_virtio.c"
+	"${CMAKE_CURRENT_LIST_DIR}/rss_comms_mhu.c"
+	)
+
+target_include_directories(${TGT}
+	 PUBLIC
+		"${CMAKE_CURRENT_LIST_DIR}"
+	)
diff --git a/components/messaging/rss_comms/sp/rss_comms_messenger.c b/components/messaging/rss_comms/sp/rss_comms_messenger.c
new file mode 100644
index 0000000..8b591f1
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_messenger.c
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <stddef.h>
+#include <trace.h>
+#include "rss_comms_messenger.h"
+#include "rss_comms_messenger_api.h"
+#include "rss_comms_mhu.h"
+#include "rss_comms_virtio.h"
+#include <protocols/rpc/common/packed-c/status.h>
+
+static const struct rss_comms_platform_ops rss_comms_ops = {
+	.transport_init = rss_comms_mhu_init,
+	.transport_deinit = rss_comms_mhu_deinit,
+	.platform_init = rss_comms_virtio_init,
+	.platform_deinit = rss_comms_virtio_deinit,
+};
+
+int rss_comms_messenger_init(struct rss_comms_messenger *rss_comms_msg)
+{
+	const struct rss_comms_platform_ops *ops;
+	int ret;
+
+	rss_comms_msg->platform_ops = &rss_comms_ops;
+	ops = rss_comms_msg->platform_ops;
+
+	ret = ops->transport_init(rss_comms_msg);
+	if (ret < 0) {
+		EMSG("[RSS-COMMS] transport init failed\n");
+		return ret;
+	}
+
+	ret = ops->platform_init(rss_comms_msg);
+	if (ret < 0) {
+		EMSG("[RSS-COMMS] platform init failed\n");
+		return ret;
+	}
+}
diff --git a/components/messaging/rss_comms/sp/rss_comms_messenger.h b/components/messaging/rss_comms/sp/rss_comms_messenger.h
new file mode 100644
index 0000000..91b5322
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_messenger.h
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __RSS_COMMS_MESSENGER_H__
+#define __RSS_COMMS_MESSENGER_H__
+
+#include <stddef.h>
+#include <stdint.h>
+#include "rss_comms_mhu.h"
+#include "rss_comms_virtio.h"
+
+struct rss_comms_platform_ops {
+	int (*transport_init)(struct rss_comms_messenger *rss_comms_msg);
+	int (*transport_deinit)(struct rss_comms_messenger *rss_comms_msg);
+	int (*platform_init)(struct rss_comms_messenger *rss_comms_msg);
+	int (*platform_deinit)(struct rss_comms_messenger *rss_comms_msg);
+};
+
+#endif /* __RSS_COMMS_MESSENGER_H__ */
diff --git a/components/messaging/rss_comms/sp/rss_comms_messenger_api.h b/components/messaging/rss_comms/sp/rss_comms_messenger_api.h
new file mode 100644
index 0000000..461d070
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_messenger_api.h
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+
+#ifndef __RSS_COMMS_MESSENGER_API_H__
+#define __RSS_COMMS_MESSENGER_API_H__
+
+#include <stddef.h>
+#include <stdint.h>
+#include "rss_comms_messenger.h"
+
+struct rss_comms_messenger {
+	const struct rss_comms_platform_ops *platform_ops;
+	void *transport;
+	void *platform;
+};
+
+int rss_comms_messenger_init(struct rss_comms_messenger *rss_comms);
+
+#endif /* __RSS_COMMS_MESSENGER_API_H__ */
diff --git a/components/messaging/rss_comms/sp/rss_comms_mhu.c b/components/messaging/rss_comms/sp/rss_comms_mhu.c
new file mode 100644
index 0000000..cff1688
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_mhu.c
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <config/interface/config_store.h>
+#include <config/interface/config_blob.h>
+#include <platform/interface/device_region.h>
+#include <platform/drivers/arm/mhu_driver/mhu_v3_x.h>
+#include <platform/drivers/arm/mhu_driver/mhu.h>
+#include <trace.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <limits.h>
+#include "rss_comms_caller.h"
+#include "rss_comms_mhu.h"
+#include "rss_comms_messenger_api.h"
+
+static int rss_comms_mhu_device_get(const char *dev,
+				  struct device_region *dev_region)
+{
+	bool found;
+
+	found = config_store_query(CONFIG_CLASSIFIER_DEVICE_REGION, dev, 0,
+				   dev_region, sizeof(*dev_region));
+	if (!found)
+		return -EINVAL;
+
+	if (!dev_region->base_addr)
+		return -EINVAL;
+
+	IMSG("mhu: device region found: %s addr: 0x%lx size: %ld", dev,
+	     dev_region->base_addr, dev_region->io_region_size);
+
+	return 0;
+}
+
+int rss_comms_mhu_init(struct rss_comms_messenger *rss_comms_msg)
+{
+	struct mhu_v3_x_dev_t *rx_dev;
+	struct mhu_v3_x_dev_t *tx_dev;
+	struct rss_comms_mhu *mhu;
+	enum mhu_error_t err;
+	int ret;
+
+	/* if we already have initialized skip this */
+	if (rss_comms_msg->transport) {
+		IMSG("[RSS-COMMS] Transport already initialized\n");
+		return 0;
+	}
+
+	mhu = malloc(sizeof(*mhu));
+	if (!mhu) {
+		EMSG("[RSS-COMMS] malloc mhu failed\n");
+		return -1;
+	}
+
+	ret = rss_comms_mhu_device_get("mhu-sender", &mhu->tx_region);
+	if (ret < 0)
+		goto free_mhu;
+
+	ret = rss_comms_mhu_device_get("mhu-receiver", &mhu->rx_region);
+	if (ret < 0)
+		goto free_mhu;
+
+	rx_dev = &mhu->rx_dev;
+	tx_dev = &mhu->tx_dev;
+
+	rx_dev->base =  (unsigned int)mhu->rx_region.base_addr;
+	rx_dev->frame = MHU_V3_X_MBX_FRAME;
+
+	tx_dev->base =  (unsigned int)mhu->tx_region.base_addr;
+	tx_dev->frame = MHU_V3_X_PBX_FRAME;
+
+	err = mhu_init_sender(tx_dev);
+	if (err != MHU_ERR_NONE) {
+		if (err == MHU_ERR_ALREADY_INIT) {
+			EMSG("[RSS-COMMS] TS to RSS MHU driver already initialized\n");
+		} else {
+			EMSG("[RSS-COMMS] TS to RSS MHU driver initialization failed: %d\n", err);
+			goto free_mhu;
+		}
+	}
+
+	err = mhu_init_receiver(rx_dev);
+	if (err != MHU_ERR_NONE) {
+		if (err == MHU_ERR_ALREADY_INIT) {
+			EMSG("[RSS-COMMS] RSS to TS MHU driver already initialized\n");
+		} else {
+			EMSG("[RSS-COMMS] RSS to TS MHU driver initialization failed: %d\n", err);
+			goto free_mhu;
+		}
+	}
+
+	rss_comms_msg->transport = (void *)mhu;
+
+	return 0;
+
+free_mhu:
+	free(mhu);
+
+	return ret;
+}
+
+int rss_comms_mhu_deinit(struct rss_comms_messenger *rss_comms_msg)
+{
+	struct rss_comms_mhu *mhu;
+
+	if (!rss_comms_msg->transport)
+		return 0;
+
+	mhu = rss_comms_msg->transport;
+	free(mhu);
+
+	rss_comms_msg->transport = NULL;
+
+	return 0;
+}
diff --git a/components/messaging/rss_comms/sp/rss_comms_mhu.h b/components/messaging/rss_comms/sp/rss_comms_mhu.h
new file mode 100644
index 0000000..f54dd8c
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_mhu.h
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __RSS_COMMS_MHU_H__
+#define __RSS_COMMS_MHU_H__
+
+#include <stddef.h>
+#include "rss_comms_caller.h"
+#include <platform/interface/device_region.h>
+#include <platform/drivers/arm/mhu_driver/mhu_v3_x.h>
+#include "rss_comms_messenger_api.h"
+
+extern struct rss_comms_messenger;
+
+struct rss_comms_mhu {
+	struct device_region rx_region;
+	struct device_region tx_region;
+	struct mhu_v3_x_dev_t rx_dev;
+	struct mhu_v3_x_dev_t tx_dev;
+};
+
+int rss_comms_mhu_init(struct rss_comms_messenger *rss_comms_msg);
+int rss_comms_mhu_deinit(struct rss_comms_messenger *rss_comms_msg);
+
+#endif /* __RSS_COMMS_MHU_H__ */
diff --git a/components/messaging/rss_comms/sp/rss_comms_virtio.c b/components/messaging/rss_comms/sp/rss_comms_virtio.c
new file mode 100644
index 0000000..664f36f
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_virtio.c
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <platform/interface/device_region.h>
+#include <config/interface/config_store.h>
+#include <config/interface/config_blob.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <limits.h>
+#include <stddef.h>
+#include <trace.h>
+#include "rss_comms_caller.h"
+#include "rss_comms_virtio.h"
+
+#define RSS_COMMS_SHEM_PHYS 0xA03F0000
+
+
+static int rss_comms_virtio_device_get(const char *dev,
+				     struct device_region *dev_region)
+{
+	bool found;
+
+	found = config_store_query(CONFIG_CLASSIFIER_DEVICE_REGION, dev, 0,
+				   dev_region, sizeof(*dev_region));
+	if (!found) {
+		EMSG("[RSS-COMMS] virtio: device region not found: %s\n", dev);
+		return -EINVAL;
+	}
+
+	if (dev_region->base_addr == 0 || dev_region->io_region_size == 0) {
+		EMSG("[RSS-COMMS] virtio: device region not valid\n");
+		return -EINVAL;
+	}
+
+	IMSG("rss_comms: virtio: device region found: %s addr: 0x%lx size: %ld\n",
+	     dev, dev_region->base_addr, dev_region->io_region_size);
+
+	return  0;
+}
+
+static void rss_comms_virtio_shm_set(struct rss_comms_virtio *virtio,
+				   struct device_region *virtio_region)
+{
+	struct rss_comms_virtio_shm *shm = &virtio->shm;
+
+	shm->base_addr = virtio_region->base_addr;
+	shm->size = virtio_region->io_region_size;
+
+	IMSG("SHEM: base: 0x%0lx size: 0x%0lx size: %ld",
+	     shm->base_addr, shm->size, shm->size);
+}
+
+int rss_comms_virtio_init(struct rss_comms_messenger *rss_comms_msg)
+{
+	struct device_region virtio_dev;
+	struct rss_comms_virtio *virtio;
+	int ret;
+
+	if (!rss_comms_msg) {
+		EMSG("[RSS-COMMS] rss comms messenger is null\n");
+		return -1;
+	}
+
+	if (rss_comms_msg->platform) {
+		IMSG("[RSS-COMMS] rss comms platform already initialized\n");
+		return 0;
+	}
+
+	virtio = malloc(sizeof(*virtio));
+	if (!virtio) {
+		EMSG("[RSS-COMMS] rss comms malloc virtio failed\n");
+		return -ENOMEM;
+	}
+
+	virtio->rss_comms_msg = rss_comms_msg;
+
+	ret = rss_comms_virtio_device_get("rss_comms-virtio", &virtio_dev);
+	if (ret < 0) {
+		EMSG("[RSS-COMMS] rss comms virtio memory mapping failed");
+		goto free_virtio;
+	}
+
+	rss_comms_virtio_shm_set(virtio, &virtio_dev);
+
+	rss_comms_msg->platform = virtio;
+
+	return 0;
+
+free_virtio:
+	free(virtio);
+
+	return ret;
+}
+
+int rss_comms_virtio_deinit(struct rss_comms_messenger *rss_comms_msg)
+{
+
+	struct rss_comms_virtio *virtio;
+
+	if (!rss_comms_msg->platform)
+		return 0;
+
+	virtio = rss_comms_msg->platform;
+
+	free(virtio);
+
+	rss_comms_msg->platform = NULL;
+
+	return 0;
+}
diff --git a/components/messaging/rss_comms/sp/rss_comms_virtio.h b/components/messaging/rss_comms/sp/rss_comms_virtio.h
new file mode 100644
index 0000000..cceb7a4
--- /dev/null
+++ b/components/messaging/rss_comms/sp/rss_comms_virtio.h
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __RSS_COMMS_VIRTIO_H__
+#define __RSS_COMMS_VIRTIO_H__
+
+#include <stddef.h>
+#include "rss_comms_caller.h"
+
+/* Union as message space and reply space are never used at the same time, and this saves space as
+ * we can overlap them.
+ */
+union __packed __attribute__((aligned(4))) rss_comms_io_buffer_t {
+	struct serialized_rss_comms_msg_t msg;
+	struct serialized_rss_comms_reply_t reply;
+};
+
+struct rss_comms_virtio_shm {
+	uintptr_t base_addr;
+	size_t size;
+};
+
+struct rss_comms_virtio {
+	struct rss_comms_messenger *rss_comms_msg;
+	struct rss_comms_virtio_shm shm;
+	union rss_comms_io_buffer_t io_buf;
+	size_t msg_size;
+};
+
+int rss_comms_virtio_init(struct rss_comms_messenger *rss_comms_msg);
+int rss_comms_virtio_deinit(struct rss_comms_messenger *rss_comms_msg);
+
+#endif /* __RSS_COMMS_VIRTIO_H__ */
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_caller.c b/components/rpc/rss_comms/caller/sp/rss_comms_caller.c
new file mode 100644
index 0000000..bc3052a
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_caller.c
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stddef.h>
+#include <trace.h>
+#include "rss_comms_caller.h"
+#include "rss_comms_mhu.h"
+#include "rss_comms_virtio.h"
+#include <protocols/rpc/common/packed-c/status.h>
+#include <platform/drivers/arm/mhu_driver/mhu_v3_x.h>
+
+static rpc_status_t rss_comms_call_invoke(void *context, rpc_call_handle handle,
+                                          uint32_t opcode, int *opstatus,
+                                          uint8_t **resp_buf, size_t *resp_len)
+{
+	struct rss_comms_caller *rss_comms = context;
+	struct mhu_v3_x_dev_t *rx_dev;
+	struct mhu_v3_x_dev_t *tx_dev;
+	struct rss_comms_mhu *mhu;
+	struct rss_comms_virtio *virtio;
+	struct rss_comms_virtio_shm *shm;
+	union rss_comms_io_buffer_t io_buf;
+	struct rss_comms_messenger *rpmsg;
+	enum mhu_error_t err;
+
+	(void)opcode;
+
+	if ((handle != rss_comms) || !opstatus || !resp_buf || !resp_len || !rss_comms) {
+		EMSG("[RSS-COMMS]: call_invoke: invalid arguments\n");
+		return TS_RPC_ERROR_INVALID_PARAMETER;
+	}
+
+	rpmsg = &rss_comms->rss_comms_msg;
+	if (!rpmsg->transport) {
+		EMSG("[RSS-COMMS] transport is null\n");
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+
+	mhu = rpmsg->transport;
+	rx_dev = &mhu->rx_dev;
+	tx_dev = &mhu->tx_dev;
+	virtio = rpmsg->platform;
+	shm = &virtio->shm;
+	io_buf = virtio->io_buf;
+
+	err = mhu_send_data(tx_dev, shm->base_addr, (uint8_t *)&io_buf.msg, virtio->msg_size);
+	if (err != MHU_ERR_NONE) {
+		EMSG("[RSS-COMMS] mhu send data failed!\n");
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+	}
+
+#if DEBUG
+	/*
+	 * Poisoning the message buffer (with a known pattern).
+	 * Helps in detecting hypothetical RSS communication bugs.
+	 */
+	memset(&io_buf.msg, 0xA5, msg_size);
+#endif
+
+	err = mhu_receive_data(rx_dev, shm->base_addr, (uint8_t *)&io_buf.reply, resp_len);
+	if (err != MHU_ERR_NONE) {
+		EMSG("[RSS-COMMS] mhu receive data failed!\n");
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+	}
+
+	*resp_buf = &io_buf.reply;
+	*opstatus = 0;
+
+	return TS_RPC_CALL_ACCEPTED;
+}
+
+struct rpc_caller *rss_comms_caller_init(struct rss_comms_caller *rss_comms)
+{
+	struct rpc_caller *rpc = &rss_comms->rpc_caller;
+	int ret;
+
+	if (!rss_comms) {
+		EMSG("[RSS-COMMS] caller: invalid argument\n");
+		return NULL;
+	}
+
+	ret = rss_comms_messenger_init(&rss_comms->rss_comms_msg);
+	if (ret < 0) {
+		EMSG("[RSS-COMMS] mssenger init failed\n");
+		return NULL;
+	}
+
+	rpc_caller_init(rpc, rss_comms);
+
+	rpc->call_begin = NULL;
+	rpc->call_invoke = rss_comms_call_invoke;
+	rpc->call_end = NULL;
+
+	return rpc;
+}
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_caller.h b/components/rpc/rss_comms/caller/sp/rss_comms_caller.h
new file mode 100644
index 0000000..2a4ac3e
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_caller.h
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __RSS_COMMS_CALLER_H__
+#define __RSS_COMMS_CALLER_H__
+
+#include <rpc_caller.h>
+#include "rss_comms_protocol.h"
+#include "rss_comms_messenger_api.h"
+
+struct rss_comms_caller {
+	struct rpc_caller rpc_caller;
+	struct rss_comms_messenger rss_comms_msg;
+};
+
+struct rpc_caller *rss_comms_caller_init(struct rss_comms_caller *rss_comms);
+void rss_comms_caller_deinit(struct rss_comms_caller *rss_comms);
+
+#endif /* __RSS_COMMS_CALLER_H__ */
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol.c b/components/rpc/rss_comms/caller/sp/rss_comms_protocol.c
new file mode 100644
index 0000000..b649779
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol.c
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#include <assert.h>
+#include "rss_comms_protocol.h"
+
+psa_status_t rss_protocol_serialize_msg(psa_handle_t handle,
+					int16_t type,
+					const psa_invec *in_vec,
+					uint8_t in_len,
+					const psa_outvec *out_vec,
+					uint8_t out_len,
+					struct serialized_rss_comms_msg_t *msg,
+					size_t *msg_len)
+{
+	psa_status_t status;
+
+	assert(msg != NULL);
+	assert(msg_len != NULL);
+	assert(in_vec != NULL);
+
+	switch (msg->header.protocol_ver) {
+	case RSS_COMMS_PROTOCOL_EMBED:
+		status = rss_protocol_embed_serialize_msg(handle, type, in_vec, in_len, out_vec,
+							  out_len, &msg->msg.embed, msg_len);
+		if (status != PSA_SUCCESS) {
+			return status;
+		}
+		break;
+	case RSS_COMMS_PROTOCOL_POINTER_ACCESS:
+		status = rss_protocol_pointer_access_serialize_msg(handle, type, in_vec, in_len,
+								   out_vec, out_len,
+								   &msg->msg.pointer_access,
+								   msg_len);
+		if (status != PSA_SUCCESS) {
+			return status;
+		}
+		break;
+	default:
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	*msg_len += sizeof(struct serialized_rss_comms_header_t);
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rss_protocol_deserialize_reply(psa_outvec *out_vec,
+					    uint8_t out_len,
+					    psa_status_t *return_val,
+					    const struct serialized_rss_comms_reply_t *reply,
+					    size_t reply_size)
+{
+	assert(reply != NULL);
+	assert(return_val != NULL);
+
+	switch (reply->header.protocol_ver) {
+	case RSS_COMMS_PROTOCOL_EMBED:
+		return rss_protocol_embed_deserialize_reply(out_vec, out_len, return_val,
+							    &reply->reply.embed, reply_size);
+	case RSS_COMMS_PROTOCOL_POINTER_ACCESS:
+		return rss_protocol_pointer_access_deserialize_reply(out_vec, out_len, return_val,
+								     &reply->reply.pointer_access,
+								     reply_size);
+	default:
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	return PSA_SUCCESS;
+}
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol.h b/components/rpc/rss_comms/caller/sp/rss_comms_protocol.h
new file mode 100644
index 0000000..116ca80
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol.h
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __RSS_COMMS_PROTOCOL_H__
+#define __RSS_COMMS_PROTOCOL_H__
+
+#include <stdint.h>
+#include <psa/client.h>
+#include <service/common/include/psa/client.h>
+#include "rss_comms_protocol_embed.h"
+#include "rss_comms_protocol_pointer_access.h"
+
+typedef struct psa_invec psa_invec;
+typedef struct psa_outvec psa_outvec;
+
+enum rss_comms_protocol_version_t {
+	RSS_COMMS_PROTOCOL_EMBED = 0,
+	RSS_COMMS_PROTOCOL_POINTER_ACCESS = 1,
+};
+
+struct __packed serialized_rss_comms_header_t {
+	uint8_t protocol_ver;
+	uint8_t seq_num;
+	uint16_t client_id;
+};
+
+/* MHU message passed from Host to RSS to deliver a PSA client call */
+struct __packed serialized_rss_comms_msg_t {
+	struct serialized_rss_comms_header_t header;
+	union __packed {
+		struct rss_embed_msg_t embed;
+		struct rss_pointer_access_msg_t pointer_access;
+	} msg;
+};
+
+/* MHU reply message to hold the PSA client reply result returned by RSS */
+struct __packed serialized_rss_comms_reply_t {
+	struct serialized_rss_comms_header_t header;
+	union __packed {
+		struct rss_embed_reply_t embed;
+		struct rss_pointer_access_reply_t pointer_access;
+	} reply;
+};
+
+/* in_len and out_len are uint8_ts, therefore if there are more than 255 iovecs
+ * an error may occur.
+ */
+//CASSERT(PSA_MAX_IOVEC <= UINT8_MAX, assert_rss_comms_max_iovec_too_large);
+
+psa_status_t rss_protocol_serialize_msg(psa_handle_t handle,
+					int16_t type,
+					const psa_invec *in_vec,
+					uint8_t in_len,
+					const psa_outvec *out_vec,
+					uint8_t out_len,
+					struct serialized_rss_comms_msg_t *msg,
+					size_t *msg_len);
+
+psa_status_t rss_protocol_deserialize_reply(psa_outvec *out_vec,
+					    uint8_t out_len,
+					    psa_status_t *return_val,
+					    const struct serialized_rss_comms_reply_t *reply,
+					    size_t reply_size);
+
+#endif /* __RSS_COMMS_PROTOCOL_H__ */
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.c b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.c
new file mode 100644
index 0000000..6d1740c
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.c
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2022-2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#include <assert.h>
+#include <string.h>
+#include "rss_comms_protocol_embed.h"
+
+#define TYPE_OFFSET	(16U)
+#define TYPE_MASK	(0xFFFFUL << TYPE_OFFSET)
+#define IN_LEN_OFFSET	(8U)
+#define IN_LEN_MASK	(0xFFUL << IN_LEN_OFFSET)
+#define OUT_LEN_OFFSET	(0U)
+#define OUT_LEN_MASK	(0xFFUL << OUT_LEN_OFFSET)
+
+#define PARAM_PACK(type, in_len, out_len)			  \
+	(((((uint32_t)type) << TYPE_OFFSET) & TYPE_MASK)	| \
+	 ((((uint32_t)in_len) << IN_LEN_OFFSET) & IN_LEN_MASK)	| \
+	 ((((uint32_t)out_len) << OUT_LEN_OFFSET) & OUT_LEN_MASK))
+
+psa_status_t rss_protocol_embed_serialize_msg(psa_handle_t handle,
+					      int16_t type,
+					      const psa_invec *in_vec,
+					      uint8_t in_len,
+					      const psa_outvec *out_vec,
+					      uint8_t out_len,
+					      struct rss_embed_msg_t *msg,
+					      size_t *msg_len)
+{
+	uint32_t payload_size = 0;
+	uint32_t i;
+
+	assert(msg != NULL);
+	assert(msg_len != NULL);
+	assert(in_vec != NULL);
+
+	msg->ctrl_param = PARAM_PACK(type, in_len, out_len);
+	msg->handle = handle;
+
+	/* Fill msg iovec lengths */
+	for (i = 0U; i < in_len; ++i) {
+		msg->io_size[i] = in_vec[i].len;
+	}
+	for (i = 0U; i < out_len; ++i) {
+		msg->io_size[in_len + i] = out_vec[i].len;
+	}
+
+	for (i = 0U; i < in_len; ++i) {
+		if (in_vec[i].len > sizeof(msg->trailer) - payload_size) {
+			return PSA_ERROR_INVALID_ARGUMENT;
+		}
+		memcpy(msg->trailer + payload_size,
+		       in_vec[i].base,
+		       in_vec[i].len);
+		payload_size += in_vec[i].len;
+	}
+
+	/* Output the actual size of the message, to optimize sending */
+	*msg_len = sizeof(*msg) - sizeof(msg->trailer) + payload_size;
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rss_protocol_embed_deserialize_reply(psa_outvec *out_vec,
+						  uint8_t out_len,
+						  psa_status_t *return_val,
+						  const struct rss_embed_reply_t *reply,
+						  size_t reply_size)
+{
+	uint32_t payload_offset = 0;
+	uint32_t i;
+
+	assert(reply != NULL);
+	assert(return_val != NULL);
+
+	for (i = 0U; i < out_len; ++i) {
+		if ((sizeof(*reply) - sizeof(reply->trailer) + payload_offset)
+		    > reply_size) {
+			return PSA_ERROR_INVALID_ARGUMENT;
+		}
+
+		memcpy(out_vec[i].base,
+		       reply->trailer + payload_offset,
+		       reply->out_size[i]);
+		out_vec[i].len = reply->out_size[i];
+		payload_offset += reply->out_size[i];
+	}
+
+	*return_val = reply->return_val;
+
+	return PSA_SUCCESS;
+}
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.h b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.h
new file mode 100644
index 0000000..1914143
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_embed.h
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2022-2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __RSS_COMMS_PROTOCOL_EMBED_H__
+#define __RSS_COMMS_PROTOCOL_EMBED_H__
+
+#include <sys/cdefs.h>
+#include <psa/client.h>
+
+#define PLAT_RSS_COMMS_PAYLOAD_MAX_SIZE 0x500
+
+typedef struct psa_invec psa_invec;
+typedef struct psa_outvec psa_outvec;
+
+struct __packed rss_embed_msg_t {
+	psa_handle_t handle;
+	uint32_t ctrl_param; /* type, in_len, out_len */
+	uint16_t io_size[PSA_MAX_IOVEC];
+	uint8_t trailer[PLAT_RSS_COMMS_PAYLOAD_MAX_SIZE];
+};
+
+struct __packed rss_embed_reply_t {
+	int32_t return_val;
+	uint16_t out_size[PSA_MAX_IOVEC];
+	uint8_t trailer[PLAT_RSS_COMMS_PAYLOAD_MAX_SIZE];
+};
+
+psa_status_t rss_protocol_embed_serialize_msg(psa_handle_t handle,
+					      int16_t type,
+					      const struct psa_invec *in_vec,
+					      uint8_t in_len,
+					      const psa_outvec *out_vec,
+					      uint8_t out_len,
+					      struct rss_embed_msg_t *msg,
+					      size_t *msg_len);
+
+psa_status_t rss_protocol_embed_deserialize_reply(psa_outvec *out_vec,
+						  uint8_t out_len,
+						  psa_status_t *return_val,
+						  const struct rss_embed_reply_t *reply,
+						  size_t reply_size);
+
+#endif /* __RSS_COMMS_PROTOCOL_EMBED_H__ */
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.c b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.c
new file mode 100644
index 0000000..32a7fd0
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.c
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022-2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#include <assert.h>
+#include "rss_comms_protocol_pointer_access.h"
+
+#define TYPE_OFFSET	(16U)
+#define TYPE_MASK	(0xFFFFUL << TYPE_OFFSET)
+#define IN_LEN_OFFSET	(8U)
+#define IN_LEN_MASK	(0xFFUL << IN_LEN_OFFSET)
+#define OUT_LEN_OFFSET	(0U)
+#define OUT_LEN_MASK	(0xFFUL << OUT_LEN_OFFSET)
+
+#define PARAM_PACK(type, in_len, out_len)			  \
+	(((((uint32_t)type) << TYPE_OFFSET) & TYPE_MASK)	| \
+	 ((((uint32_t)in_len) << IN_LEN_OFFSET) & IN_LEN_MASK)	| \
+	 ((((uint32_t)out_len) << OUT_LEN_OFFSET) & OUT_LEN_MASK))
+
+psa_status_t rss_protocol_pointer_access_serialize_msg(psa_handle_t handle,
+						       int16_t type,
+						       const psa_invec *in_vec,
+						       uint8_t in_len,
+						       const psa_outvec *out_vec,
+						       uint8_t out_len,
+						       struct rss_pointer_access_msg_t *msg,
+						       size_t *msg_len)
+{
+	unsigned int i;
+
+	assert(msg != NULL);
+	assert(msg_len != NULL);
+	assert(in_vec != NULL);
+
+	msg->ctrl_param = PARAM_PACK(type, in_len, out_len);
+	msg->handle = handle;
+
+	/* Fill msg iovec lengths */
+	for (i = 0U; i < in_len; ++i) {
+		msg->io_sizes[i] = in_vec[i].len;
+		msg->host_ptrs[i] = (uint64_t)in_vec[i].base;
+	}
+	for (i = 0U; i < out_len; ++i) {
+		msg->io_sizes[in_len + i] = out_vec[i].len;
+		msg->host_ptrs[in_len + i] = (uint64_t)out_vec[i].base;
+	}
+
+	*msg_len = sizeof(*msg);
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rss_protocol_pointer_access_deserialize_reply(psa_outvec *out_vec,
+							   uint8_t out_len,
+							   psa_status_t *return_val,
+							   const struct rss_pointer_access_reply_t *reply,
+							   size_t reply_size)
+{
+	unsigned int i;
+
+	assert(reply != NULL);
+	assert(return_val != NULL);
+
+	for (i = 0U; i < out_len; ++i) {
+		out_vec[i].len = reply->out_sizes[i];
+	}
+
+	*return_val = reply->return_val;
+
+	return PSA_SUCCESS;
+}
diff --git a/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.h b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.h
new file mode 100644
index 0000000..3f09d65
--- /dev/null
+++ b/components/rpc/rss_comms/caller/sp/rss_comms_protocol_pointer_access.h
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2022-2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __RSS_COMMS_PROTOCOL_POINTER_ACCESS_H__
+#define __RSS_COMMS_PROTOCOL_POINTER_ACCESS_H__
+
+#include <sys/cdefs.h>
+#include <psa/client.h>
+
+typedef struct psa_invec psa_invec;
+typedef struct psa_outvec psa_outvec;
+
+struct __packed rss_pointer_access_msg_t {
+	psa_handle_t handle;
+	uint32_t ctrl_param;
+	uint32_t io_sizes[PSA_MAX_IOVEC];
+	uint64_t host_ptrs[PSA_MAX_IOVEC];
+};
+
+struct __packed rss_pointer_access_reply_t {
+	int32_t return_val;
+	uint32_t out_sizes[PSA_MAX_IOVEC];
+};
+
+psa_status_t rss_protocol_pointer_access_serialize_msg(psa_handle_t handle,
+						       int16_t type,
+						       const psa_invec *in_vec,
+						       uint8_t in_len,
+						       const psa_outvec *out_vec,
+						       uint8_t out_len,
+						       struct rss_pointer_access_msg_t *msg,
+						       size_t *msg_len);
+
+psa_status_t rss_protocol_pointer_access_deserialize_reply(psa_outvec *out_vec,
+							   uint8_t out_len,
+							   psa_status_t *return_val,
+							   const struct rss_pointer_access_reply_t *reply,
+							   size_t reply_size);
+
+#endif /* __RSS_COMMS_PROTOCOL_POINTER_ACCESS_H__ */
diff --git a/components/rpc/rss_comms/component.cmake b/components/rpc/rss_comms/component.cmake
new file mode 100644
index 0000000..ca5eea5
--- /dev/null
+++ b/components/rpc/rss_comms/component.cmake
@@ -0,0 +1,27 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2021-2023, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+if (NOT DEFINED TGT)
+	message(FATAL_ERROR "mandatory parameter TGT is not defined.")
+endif()
+
+set_property(TARGET ${TGT} APPEND PROPERTY PUBLIC_HEADER
+	${CMAKE_CURRENT_LIST_DIR}/caller/sp/rss_comms_caller.h
+	)
+
+target_sources(${TGT} PRIVATE
+	"${CMAKE_CURRENT_LIST_DIR}/service_rss_comms.c"
+	"${CMAKE_CURRENT_LIST_DIR}/caller/sp/rss_comms_caller.c"
+	"${CMAKE_CURRENT_LIST_DIR}/caller/sp/rss_comms_protocol.c"
+	"${CMAKE_CURRENT_LIST_DIR}/caller/sp/rss_comms_protocol_embed.c"
+	"${CMAKE_CURRENT_LIST_DIR}/caller/sp/rss_comms_protocol_pointer_access.c"
+	)
+
+target_include_directories(${TGT}
+	 PUBLIC
+		"${CMAKE_CURRENT_LIST_DIR}/caller/sp/"
+	)
+
diff --git a/components/rpc/rss_comms/service_rss_comms.c b/components/rpc/rss_comms/service_rss_comms.c
new file mode 100644
index 0000000..b90f697
--- /dev/null
+++ b/components/rpc/rss_comms/service_rss_comms.c
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2021-2023, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2021-2023, Linaro Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <platform/drivers/arm/mhu_driver/mhu_v3_x.h>
+#include <psa/client.h>
+#include <rss_comms_protocol.h>
+#include "rss_comms_caller.h"
+#include "trace.h"
+#include "rss_comms_messenger_api.h"
+#include <protocols/rpc/common/packed-c/status.h>
+
+#define AP_RSS_SND_MHU_BASE    0x2AB20000
+#define AP_RSS_RCV_MHU_BASE    0x2AB30000
+
+extern size_t mhu_get_max_message_size(void);
+
+static uint8_t select_protocol_version(const psa_invec *in_vec, size_t in_len,
+				       const psa_outvec *out_vec, size_t out_len)
+{
+	size_t comms_mhu_msg_size;
+	size_t comms_embed_msg_min_size;
+	size_t comms_embed_reply_min_size;
+	size_t in_size_total = 0;
+	size_t out_size_total = 0;
+	size_t i;
+
+	for (i = 0U; i < in_len; ++i) {
+		in_size_total += in_vec[i].len;
+	}
+	for (i = 0U; i < out_len; ++i) {
+		out_size_total += out_vec[i].len;
+	}
+
+	comms_mhu_msg_size = mhu_get_max_message_size();
+
+	comms_embed_msg_min_size = sizeof(struct serialized_rss_comms_header_t) +
+				   sizeof(struct rss_embed_msg_t) -
+				   PLAT_RSS_COMMS_PAYLOAD_MAX_SIZE;
+
+	comms_embed_reply_min_size = sizeof(struct serialized_rss_comms_header_t) +
+				     sizeof(struct rss_embed_reply_t) -
+				     PLAT_RSS_COMMS_PAYLOAD_MAX_SIZE;
+
+	/* Use embed if we can pack into one message and reply, else use
+	 * pointer_access. The underlying MHU transport protocol uses a
+	 * single uint32_t to track the length, so the amount of data that
+	 * can be in a message is 4 bytes less than mhu_get_max_message_size
+	 * reports.
+	 *
+	 * TODO tune this with real performance numbers, it's possible a
+	 * pointer_access message is less performant than multiple embed
+	 * messages due to ATU configuration costs to allow access to the
+	 * pointers.
+	 */
+	if ((comms_embed_msg_min_size + in_size_total > comms_mhu_msg_size - sizeof(uint32_t))
+	 || (comms_embed_reply_min_size + out_size_total > comms_mhu_msg_size - sizeof(uint32_t))) {
+		return RSS_COMMS_PROTOCOL_POINTER_ACCESS;
+	} else {
+		return RSS_COMMS_PROTOCOL_EMBED;
+	}
+}
+
+psa_status_t __psa_call(struct rpc_caller *caller, psa_handle_t handle,
+		               int32_t client_id, int32_t type,
+					   const struct psa_invec *in_vec, size_t in_len,
+		               struct psa_outvec *out_vec, size_t out_len)
+{
+	/* Declared statically to avoid using huge amounts of stack space. Maybe revisit if
+	 * functions not being reentrant becomes a problem.
+	 */
+	union rss_comms_io_buffer_t io_buf;
+	struct rss_comms_messenger *rpmsg;
+	psa_status_t psa_status = PSA_SUCCESS;
+	static uint8_t seq_num = 1U;
+	psa_status_t status;
+	size_t reply_size = sizeof(io_buf.reply);
+	psa_status_t return_val;
+	struct rss_comms_virtio *virtio;
+	struct rss_comms_caller *rss_comms;
+	rpc_call_handle rpc_handle;
+	rpc_handle = caller;
+	size_t msg_size = 0;
+	int ret = 0;
+	uint8_t *resp;
+
+	if (!caller) {
+		EMSG("[RSS-COMMS] psa_call rpc caller is null\n");
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+
+	rss_comms = caller->context;
+	if( !rss_comms ) {
+		EMSG("[RSS-COMMS] psa_call rss comms is null\n");
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+
+	rpmsg = &rss_comms->rss_comms_msg;
+	if (!rpmsg->transport) {
+		EMSG("[RSS-COMMS] psa_call messenger is null\n");
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+
+	virtio = (struct rss_comms_virtio *)rpmsg->platform;
+
+	if (type > INT16_MAX || type < INT16_MIN || in_len > PSA_MAX_IOVEC
+	    || out_len > PSA_MAX_IOVEC) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	virtio->io_buf.msg.header.seq_num = seq_num;
+	/* No need to distinguish callers (currently concurrent calls are not supported). */
+	virtio->io_buf.msg.header.client_id = client_id;
+	virtio->io_buf.msg.header.protocol_ver = select_protocol_version(in_vec, in_len, out_vec, out_len);
+
+	status = rss_protocol_serialize_msg(handle, type, in_vec, in_len, out_vec,
+					    out_len, &virtio->io_buf.msg, &msg_size);
+	if (status != PSA_SUCCESS) {
+		EMSG("psa_call: serialize msg failed: %d\n", status);
+		return status;
+	}
+
+	virtio->msg_size = msg_size;
+
+	IMSG("[RSS-COMMS] Sending message\n");
+	IMSG("protocol_ver=%u\n", virtio->io_buf.msg.header.protocol_ver);
+	IMSG("seq_num=%u\n", virtio->io_buf.msg.header.seq_num);
+	IMSG("client_id=%u\n", virtio->io_buf.msg.header.client_id);
+
+	ret = rpc_caller_invoke(caller, rpc_handle, 0, &psa_status, &resp,  &reply_size);
+	if (ret != TS_RPC_CALL_ACCEPTED) {
+		EMSG("psa_call: invoke failed: %d\n", ret);
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+
+	virtio->io_buf.reply = *(struct serialized_rss_comms_reply_t *)resp;
+
+	IMSG("[RSS-COMMS] Received reply\n");
+	IMSG("protocol_ver=%u\n", virtio->io_buf.reply.header.protocol_ver);
+	IMSG("seq_num=%u\n", virtio->io_buf.reply.header.seq_num);
+	IMSG("client_id=%u\n", virtio->io_buf.reply.header.client_id);
+	IMSG("reply_size=%lu\n", reply_size);
+
+	status = rss_protocol_deserialize_reply(out_vec, out_len, &return_val,
+						&virtio->io_buf.reply, reply_size);
+	if (status != PSA_SUCCESS) {
+		EMSG("psa_call: protocol deserialize reply failed: %d\n", status);
+		return status;
+	}
+
+	IMSG("return_val=%d\n", return_val);
+
+	/* Clear the MHU message buffer to remove assets from memory */
+	memset(&virtio->io_buf, 0x0, sizeof(virtio->io_buf));
+
+	seq_num++;
+
+	return return_val;
+}
+
+psa_status_t psa_call_client_id(struct rpc_caller *caller,
+				psa_handle_t psa_handle, int32_t client_id,
+				int32_t type, const struct psa_invec *in_vec,
+				size_t in_len, struct psa_outvec *out_vec,
+				size_t out_len)
+{
+	return __psa_call(caller, psa_handle, client_id, type, in_vec, in_len,
+			  out_vec, out_len);
+}
+
+psa_status_t psa_call(struct rpc_caller *caller, psa_handle_t psa_handle,
+		      int32_t type, const struct psa_invec *in_vec,
+		      size_t in_len, struct psa_outvec *out_vec, size_t out_len)
+{
+	return __psa_call(caller, psa_handle, 0, type, in_vec, in_len, out_vec,
+			  out_len);
+}
diff --git a/components/service/secure_storage/backend/secure_storage_ipc/secure_storage_ipc.c b/components/service/secure_storage/backend/secure_storage_ipc/secure_storage_ipc.c
index 0609326..c342465 100644
--- a/components/service/secure_storage/backend/secure_storage_ipc/secure_storage_ipc.c
+++ b/components/service/secure_storage/backend/secure_storage_ipc/secure_storage_ipc.c
@@ -155,12 +155,17 @@ static uint32_t secure_storage_get_support(void *context, uint32_t client_id)
 	struct rpc_caller *caller = ipc->client.caller;
 	psa_status_t psa_status;
 	uint32_t support_flags;
+	uint32_t dummy_vec = 0;
+	/* Use dummy in vector to guarantee parameter checking */
+	struct psa_invec in_vec[] = {
+		{ .base = &dummy_vec, .len = sizeof(dummy_vec) },
+	};
 	struct psa_outvec out_vec[] = {
 		{ .base = psa_ptr_to_u32(&support_flags), .len =  sizeof(support_flags) },
 	};
 
 	psa_status = psa_call_client_id(caller, ipc->service_handle, client_id,
-					TFM_PS_ITS_GET_SUPPORT, NULL, 0,
+					TFM_PS_ITS_GET_SUPPORT, in_vec, IOVEC_LEN(in_vec),
 					out_vec, IOVEC_LEN(out_vec));
 	if (psa_status != PSA_SUCCESS)
 		EMSG("ipc_get_support: failed to psa_call: %d", psa_status);
diff --git a/deployments/se-proxy/config/kronos-opteesp/default_se-proxy.dts.in b/deployments/se-proxy/config/kronos-opteesp/default_se-proxy.dts.in
index 5ca6349..5f2afd1 100644
--- a/deployments/se-proxy/config/kronos-opteesp/default_se-proxy.dts.in
+++ b/deployments/se-proxy/config/kronos-opteesp/default_se-proxy.dts.in
@@ -21,5 +21,23 @@
 
 	device-regions {
 		compatible = "arm,ffa-manifest-device-regions";
+		mhu-sender {
+			/* Armv8 A Foundation Platform values */
+			base-address = <0x00000000 0x2AB20000>; /* AP2RSS Secure MHU Sender */
+			pages-count = <16>;
+			attributes = <0x3>; /* read-write */
+		};
+		mhu-receiver {
+			/* Armv8 A Foundation Platform values */
+			base-address = <0x00000000 0x2AB30000>; /* AP2RSS Secure MHU Receiver */
+			pages-count = <16>;
+			attributes = <0x3>; /* read-write */
+		};
+		rss_comms-virtio {
+			/* Armv8 A Foundation Platform values */
+			base-address = <0x00000000 0xA03F0000>; /* shared memory for the communication between TS and RSS */
+			pages-count = <16>;
+			attributes = <0x3>; /* read-write */
+		};
 	};
 };
diff --git a/deployments/se-proxy/infra/kronos/infra.cmake b/deployments/se-proxy/infra/kronos/infra.cmake
index 7670af4..084ce06 100644
--- a/deployments/se-proxy/infra/kronos/infra.cmake
+++ b/deployments/se-proxy/infra/kronos/infra.cmake
@@ -14,7 +14,16 @@
 add_components(TARGET "se-proxy"
 	BASE_DIR ${TS_ROOT}
 	COMPONENTS
+	"components/rpc/dummy"
 	"components/rpc/common/caller"
+	"components/rpc/rss_comms"
+	"components/messaging/rss_comms/sp"
+	"components/service/secure_storage/backend/secure_storage_ipc"
+	"components/service/attestation/reporter/stub"
+	"components/service/attestation/key_mngr/stub"
+	"components/service/crypto/backend/stub"
+	"components/service/crypto/client/psa"
+	"components/service/secure_storage/backend/mock_store"
 )
 
 target_sources(se-proxy PRIVATE
diff --git a/deployments/se-proxy/infra/kronos/service_proxy_factory.c b/deployments/se-proxy/infra/kronos/service_proxy_factory.c
index da326cc..7524c87 100644
--- a/deployments/se-proxy/infra/kronos/service_proxy_factory.c
+++ b/deployments/se-proxy/infra/kronos/service_proxy_factory.c
@@ -8,22 +8,47 @@
 #include <stddef.h>
 #include <psa/sid.h>
 #include <rpc/common/endpoint/rpc_interface.h>
+#include <rpc/rss_comms/caller/sp/rss_comms_caller.h>
 #include <service/attestation/provider/attest_provider.h>
 #include <service/attestation/provider/serializer/packed-c/packedc_attest_provider_serializer.h>
 #include <service/crypto/factory/crypto_provider_factory.h>
 #include <service/secure_storage/frontend/secure_storage_provider/secure_storage_provider.h>
 #include <trace.h>
 
+struct rss_comms_caller rss_comms;
+
+/* IPC backends */
+#include <service/secure_storage/backend/secure_storage_ipc/secure_storage_ipc.h>
+/* Stub backends */
+#include <service/crypto/backend/stub/stub_crypto_backend.h>
+#include <service/secure_storage/backend/mock_store/mock_store.h>
+
 struct rpc_interface *attest_proxy_create(void)
 {
 	struct rpc_interface *attest_iface = NULL;
 
+	/* Static objects for proxy instance */
+	static struct attest_provider attest_provider;
+
+	/* Initialize the service provider */
+	attest_iface = attest_provider_init(&attest_provider);
+
+	attest_provider_register_serializer(&attest_provider,
+		TS_RPC_ENCODING_PACKED_C, packedc_attest_provider_serializer_instance());
+
 	return attest_iface;
 }
 
 struct rpc_interface *crypto_proxy_create(void)
 {
 	struct rpc_interface *crypto_iface = NULL;
+	struct crypto_provider *crypto_provider;
+
+	if (stub_crypto_backend_init() == PSA_SUCCESS) {
+
+		crypto_provider = crypto_provider_factory_create();
+		crypto_iface = service_provider_get_rpc_interface(&crypto_provider->base_provider);
+	}
 
 	return crypto_iface;
 }
@@ -32,12 +57,30 @@ struct rpc_interface *ps_proxy_create(void)
 {
 	struct rpc_interface *rpc_interface = NULL;
 
+	static struct secure_storage_provider ps_provider;
+	static struct secure_storage_ipc ps_backend;
+	struct rpc_caller *storage_caller;
+	struct storage_backend *backend;
+
+	storage_caller = rss_comms_caller_init(&rss_comms);
+	if (!storage_caller) {
+		EMSG("storage caller is null\n");
+		return NULL;
+	}
+	backend = secure_storage_ipc_init(&ps_backend, &rss_comms.rpc_caller);
+	ps_backend.service_handle = TFM_PROTECTED_STORAGE_SERVICE_HANDLE;
+
+	rpc_interface = secure_storage_provider_init(&ps_provider, backend);
+
 	return rpc_interface;
 }
 
 struct rpc_interface *its_proxy_create(void)
 {
-	struct rpc_interface *rpc_interface = NULL;
+	static struct mock_store its_backend;
+	static struct secure_storage_provider its_provider;
 
-	return rpc_interface;
+	struct storage_backend *backend = mock_store_init(&its_backend);
+
+	return secure_storage_provider_init(&its_provider, backend);
 }
diff --git a/platform/drivers/arm/mhu_driver/mhu.h b/platform/drivers/arm/mhu_driver/mhu.h
index e9f407b..8e1b1d5 100644
--- a/platform/drivers/arm/mhu_driver/mhu.h
+++ b/platform/drivers/arm/mhu_driver/mhu.h
@@ -97,6 +97,7 @@ enum mhu_error_t mhu_send_data(void *mhu_sender_dev,
  *       multiple of 4.
  */
 enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
+                                  void *mhu_outband_base,
                                   uint8_t *receive_buffer,
                                   size_t *size);
 
diff --git a/platform/drivers/arm/mhu_driver/mhu_v3_x.c b/platform/drivers/arm/mhu_driver/mhu_v3_x.c
index b640b3b..4ee6ba2 100644
--- a/platform/drivers/arm/mhu_driver/mhu_v3_x.c
+++ b/platform/drivers/arm/mhu_driver/mhu_v3_x.c
@@ -290,7 +290,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_clear(const struct mhu_v3_x_dev_t *dev,
 {
     union _mhu_v3_x_frame_t *p_mhu;
     struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
-    enum mhu_v3_x_error_t status;
 
     if (dev == NULL) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -321,7 +320,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_write(const struct mhu_v3_x_dev_t *dev,
 {
     union _mhu_v3_x_frame_t *p_mhu;
     struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
-    enum mhu_v3_x_error_t status;
 
     if (dev == NULL) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -352,7 +350,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_read(const struct mhu_v3_x_dev_t *dev,
      uint32_t channel, uint32_t *value)
 {
     union _mhu_v3_x_frame_t *p_mhu;
-    enum mhu_v3_x_error_t status;
 
     if ((dev == NULL) || (value == NULL)) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -389,7 +386,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_set(
 {
     union _mhu_v3_x_frame_t *p_mhu;
     struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
-    enum mhu_v3_x_error_t status;
 
     if (dev == NULL) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -421,7 +417,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_clear(
 {
     union _mhu_v3_x_frame_t *p_mhu;
     struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
-    enum mhu_v3_x_error_t status;
 
     if (dev == NULL) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -453,7 +448,6 @@ enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_get(
 {
     union _mhu_v3_x_frame_t *p_mhu;
     struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
-    enum mhu_v3_x_error_t status;
 
     if ((dev == NULL) || (mask_status == NULL)) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -520,7 +514,6 @@ enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_enable(
      const struct mhu_v3_x_dev_t *dev, uint32_t channel,
      enum mhu_v3_x_channel_type_t ch_type)
 {
-    enum mhu_v3_x_error_t status;
 
     if (dev == NULL) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
@@ -582,7 +575,6 @@ enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_disable(
      const struct mhu_v3_x_dev_t *dev, uint32_t channel,
      enum mhu_v3_x_channel_type_t ch_type)
 {
-    enum mhu_v3_x_error_t status;
 
     /* Check if driver has been initialized */
     if (!(dev->is_initialized)) {
@@ -618,8 +610,6 @@ enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_clear(
      const struct mhu_v3_x_dev_t *dev, uint32_t channel,
      enum mhu_v3_x_channel_type_t ch_type)
 {
-    union _mhu_v3_x_frame_t *p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
-    enum mhu_v3_x_error_t status;
 
     /* Check if driver has been initialized */
     if (!(dev->is_initialized)) {
diff --git a/platform/drivers/arm/mhu_driver/mhu_v3_x.h b/platform/drivers/arm/mhu_driver/mhu_v3_x.h
index d14946c..18bd910 100644
--- a/platform/drivers/arm/mhu_driver/mhu_v3_x.h
+++ b/platform/drivers/arm/mhu_driver/mhu_v3_x.h
@@ -25,7 +25,6 @@
 #include <stdint.h>
 #include <stdbool.h>
 #include "mhu.h"
-//#include "host_base_address.h"
 
 #ifdef __cplusplus
 extern "C" {
@@ -251,7 +250,6 @@ enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_clear(
      const struct mhu_v3_x_dev_t *dev, uint32_t channel,
      enum mhu_v3_x_channel_type_t ch_type);
 
-
 #ifdef __cplusplus
 }
 #endif
diff --git a/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c b/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c
index 74eddec..60203bf 100644
--- a/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c
+++ b/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c
@@ -15,6 +15,7 @@
  */
 
 #include "mhu_v3_x.h"
+#include "mhu.h"
 #include <string.h>
 
 static enum mhu_error_t error_mapping_to_mhu_error_t(enum mhu_v3_x_error_t err)
@@ -236,12 +237,13 @@ enum mhu_error_t mhu_send_data(void *mhu_sender_dev, void *mhu_outband_base,
      * First 4 bytes represents the header which currently stores the size
      * of msg. Rest of buffer holds the actual message.
      */
-    /* (void)memcpy((void *)MHU3_OUTBAND_BUF_BASE, (void *)&msg_len,
-            MHU3_OUTBAND_BUF_HEADER_SIZE); */
+    (void)memcpy((void *)mhu_outband_base, (void *)&msg_len, MHU3_OUTBAND_BUF_HEADER_SIZE);
 
     /* Copy the message */
-    /* (void)memcpy((void *)(MHU3_OUTBAND_BUF_BASE + MHU3_OUTBAND_BUF_HEADER_SIZE),
-            (void *)send_buffer, msg_len); */
+    for (int i = 0; i < msg_len; i ++) {
+        (void)memcpy((void *)(mhu_outband_base + MHU3_OUTBAND_BUF_HEADER_SIZE + i),
+                     (uint8_t *)(send_buffer + i), 1);
+    }
 
     /*
      * Use only one channel 0 to ring the doorbell on receiver for
@@ -262,7 +264,7 @@ fail:
     return error_mapping_to_mhu_error_t(err);
 }
 
-enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
+enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev, void *mhu_outband_base,
         uint8_t *receive_buffer, size_t *size)
 {
     struct mhu_v3_x_dev_t *dev = mhu_receiver_dev;
@@ -288,17 +290,18 @@ enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
      * First 4 bytes represents the header which currently stores the size
      * of msg. Rest of buffer holds the actual message.
      */
-    /* (void)memcpy((void *)&msg_len, (void *)MHU3_OUTBAND_BUF_BASE,
-            MHU3_OUTBAND_BUF_HEADER_SIZE);
+    (void)memcpy((void *)&msg_len, (void *)mhu_outband_base, MHU3_OUTBAND_BUF_HEADER_SIZE);
     if (*size < msg_len) {
         return MHU_V_3_X_ERR_UNSUPPORTED;
-    } */
+    }
 
     *size = msg_len;
 
     /* Copy the message */
-    /* (void)memcpy((void *)receive_buffer, (void *)(MHU3_OUTBAND_BUF_BASE +
-                MHU3_OUTBAND_BUF_HEADER_SIZE), msg_len); */
+    for (int i = 0; i < msg_len; i ++) {
+        (void)memcpy((void *)(receive_buffer + i),
+                     (void *)(mhu_outband_base + MHU3_OUTBAND_BUF_HEADER_SIZE + i), 1);
+    }
 
     return MHU_V_3_X_ERR_NONE;
 
-- 
2.25.1

